aten/src/ATen/SparseTensorImpl.cpp

```diff
@@ -0,0 +1,70 @@
+#include <ATen/ATen.h>
+#include <ATen/SparseTensorImpl.h>
+
+namespace at {
+
+// SparseTensorImpl defaults to a [0] size tensor with one sparse dimension and
+// no dense dimensions.  This is kind of arbitrary but we want an empty tensor
+// to have one dimension (zero dimensions means its scalar), and we have
+// the invariant that dim == dimI + dimV, so *one* of the inner dimensions
+// has to be one.
+SparseTensorImpl::SparseTensorImpl(Type * type)
+    : TensorImpl(type)
+    , size_{0}
+    , dimI_(1)
+    , dimV_(0)
+    , indices_(type->toDense().toScalarType(ScalarType::Long).tensor())
+    , values_(type->toDense().tensor()) {
+      AT_ASSERT(type->is_sparse());
+    }
+
+const char * SparseTensorImpl::toString() const {
+  // TODO: also give back type information
+  return "SparseTensor";
+}
+IntList SparseTensorImpl::sizes() const {
+  return size_;
+}
+IntList SparseTensorImpl::strides() const {
+  AT_ERROR("sparse tensors do not have strides");
+}
+int64_t SparseTensorImpl::dim() const {
+  return dimI_ + dimV_;
+}
+Scalar SparseTensorImpl::localScalar() {
+  AT_ERROR("sparse tensors cannot be scalars");
```

Comment:
I don't think this is right; the requirement in dense is that numel() == 1, i.e. this is basically a conversion.  We should do the same thing here.

Reply:
It appears to be broken in master (I see localScalar is used in `is_nonzero`):

```
>>> import torch
>>> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
>>> v = torch.FloatTensor([3,      4,      5    ])
>>> y = torch.sparse.FloatTensor(i.t(), v)
>>> y.is_nonzero()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: numel is not implemented for type torch.sparse.FloatTensor
```

but seems reasonable enough to fix.

Reply:
Fixed a lot of it; while working on it I realized that "scalar" sparse tensors actually make sense as a coherent concept, but I didn't go far enough to fix code to handle this correctly.